### ISSUE TITLE
Support running with Java 21

### ISSUE THREAD

**Issue by octylFractal** [gradle/gradle#25574]
Note: Compiling and testing on Java 21 will be supported as part of Gradle 8.4 using [toolchains](https://docs.gradle.org/current/userguide/toolchains.html). This issue is only for being able to run Gradle itself on Java 21.

```[tasklist]
### Steps to complete
- [ ] https://github.com/gradle/gradle/issues/25573
- [x] Upgrade Groovy (`GroovyCoverage` as well)
- [ ] https://github.com/gradle/gradle/issues/26764
- [ ] https://github.com/gradle/gradle/pull/26913
```

**Comment by Desoroxxx** [comment]
Any ETA?

**Comment by octylFractal** [comment]
Still waiting on Kotlin 1.9.20 for the necessary version support in `JvmTarget`.

**Comment by Desoroxxx** [comment]
> Still waiting on Kotlin 1.9.20 for the necessary version support in `JvmTarget`.

I remember reading somewhere that the next Kotlin version was going to be released in December did they mean 1.10/2.0 or did they mean, 1.9.20, because waiting until December for Java 21 would be long, I am so hyped 😅

**Comment by ice1000** [comment]
I asked: https://youtrack.jetbrains.com/issue/KT-58987/Use-some-available-JVM-target-if-theres-no-JvmTarget-for-the-inferred-toolchain-version

**Comment by hb-0** [comment]
As a compilation tool, Gradle should not rely on Kotlin's support for JDK 21, but should at least provide support through Groovy

**Comment by ice1000** [comment]
> As a compilation tool, Gradle should not rely on Kotlin's support for JDK 21, but should at least provide support through Groovy

The truth is, as long as Kotlin is there, it will throw exception to prevent Gradle from running on JDK 21.

**Comment by goafabric** [comment]
On Macos im Combination with the Exec Command there is currently also the issue,
that the executable has to be fully qualified (e.g. /usr/local/bin/docker instead of "docker")
and jib will complain that it cannot pickup the credentials from a docker login.
On Linux everything seems to be fine.

a detailed descripted can be found here
https://github.com/gradle/gradle/issues/25574

**Comment by octylFractal** [comment]
> As a compilation tool, Gradle should not rely on Kotlin's support for JDK 21, but should at least provide support through Groovy

Gradle uses Kotlin to run buildscripts written in the Kotlin DSL. It is necessary for Kotlin to have support for this in order to claim that Gradle can be _run_ on Java 21. If all you need is compilation, that support is already landing in 8.4, see #25573.

**Comment by ilyas2016** [comment]
Why gradle is always to late to release the latest version support. I am already running jdk 21 with maven since day one Sep 19. switching project form maven to gradle was not a good idea.

**Comment by cocorossello** [comment]
Just try it, it works for most of the use cases. We'been compiling and running tests with jdk21 since 8.3RC1, no problem at all

**Comment by aumaOne** [comment]
Silly question from me, but as far as we're using Groovy DSL only, isn't possible to deactivate the kotlin part and run 8.3 on JDK21?

**Comment by ice1000** [comment]
Is using Kotlin beta an option for gradle?

**Comment by ijuma** [comment]
Apache Kafka uses the Groovy DSL and we are able to compile and run with Java 21 using Gradle 8.3 - in fact this is enabled and passing in our CI right now.

**Comment by ice1000** [comment]
> Apache Kafka uses the Groovy DSL and we are able to compile and run with Java 21 using Gradle 8.3 - in fact this is enabled and passing in our CI right now.

I think it's because #25573 is closed as fixed :)

**Comment by pursue503** [comment]
I have checked, and it is possible to build with JDK 21 using Gradle 8.4-rc-1.

https://docs.gradle.org/8.4-rc-1/release-notes.html

**Comment by octylFractal** [comment]
> Silly question from me, but as far as we're using Groovy DSL only, isn't possible to deactivate the kotlin part and run 8.3 on JDK21?

In theory this should work without any changes to Gradle, i.e. as long as you don't have any Kotlin buildscripts, it should work as-is.

**Comment by softprops** [comment]
It seems like a big miss that a tool that many java projects depend on are dependent koltin compilation support before we can see java compilation support. 

Regarding the kotlinscript comments, how is compiling the sources of your project's builds coupled to the sources of your built itself? I would think compiling the sources of your project would only depend on a lib for containing a version of javac. 

Or is the issue here a release to build a project with only java sources is dependent on also supporting compiling kotlin project sources mixed koltin and java sources, similar to mixed scala support?

**Comment by octylFractal** [comment]
> Regarding the kotlinscript comments, how is compiling the sources of your project's builds coupled to the sources of your built itself? I would think compiling the sources of your project would only depend on a lib for containing a version of javac.

It isn't, you can (and should) use [toolchains](https://docs.gradle.org/current/userguide/toolchains.html). This is why we track compilation and testing separately. I previously stated this in https://github.com/gradle/gradle/issues/25574#issuecomment-1732842944.

**Comment by StuAtGit** [comment]
>It seems like a big miss that a tool that many java projects depend on are dependent koltin compilation support before we can see java compilation support.

I've never seen gradle seamlessly support the latest java without a new release of gradle. At this stage, those projects should know what they're getting into. There are other options.

Even once gradle releases, they'll be projects (looking at you spring) that force you to use old versions of gradle, so you're going to have to wait a while for those to update.

TLDR; If projects want seamless forward compatibility with future versions of Java, they should use Maven. Otherwise, commit to patiently waiting  ¯\_(ツ)_/¯

Please, folks, correct me if I'm wrong here. But, my experience as always been that gradle doesn't do forward compatibility.... I'm stuck with gradle, so I'd love to be wrong. Is there a way to use Java 21 with Gradle 7.x ?

**Comment by ice1000** [comment]
Can Gradle consider using a beta version of Kotlin? https://github.com/JetBrains/kotlin/releases/tag/v1.9.20-Beta2 this is out for a week and it should probably be okay if it's just the RC/Beta version of Gradle that depends on Kotlin Beta, and replace with Kotlin release once it's published.

**Comment by ice1000** [comment]
> TLDR; If projects want seamless forward compatibility with future versions of Java, they should use Maven.

I'll never use Maven. They don't even have a working nexus publication plugin for modern versions of Java. I've wasted hours on Maven and will never look at Maven ever again.

To me, this is all Kotlin's fault. They release slowly and passively fix bugs related to things they do not use internally.

**Comment by sgammon** [comment]
We are solving this by running on JVM20 and building against JVM21, which works fine on `8.4-rc-1`, and you can always build Gradle from trunk, which we did before release 👍🏻

**Comment by goafabric** [comment]
jib + exec problems on macos are still present in 8.4 RC2
@goafabric: remindme

**Comment by ipsi-apant** [comment]
> I have confirmed that it is possible to build JDK 21 with Gradle 8.4-rc-1. Even though it's a release candidate version, it can be used if necessary.
> 
> https://docs.gradle.org/8.4-rc-1/release-notes.html


Just tried this approach (using RC-2), unfortunately it is failing

```
Unsupported Java. 
Your build is currently configured to use Java 21 and Gradle 8.4-rc-2.

Possible solution:
 - Use Java 20 as Gradle JVM: Open Gradle settings 
 - Open Gradle wrapper settings, change `distributionUrl` property to use compatible Gradle version and reload the project
```

**Comment by steve-todorov** [comment]
@ipsi-apant you probably have configured IntelliJ to run Gradle with the same JDK as the Project.
![image](https://github.com/gradle/gradle/assets/2303663/354a0d4c-f3a7-40fa-8be3-e149cc358c47)

**Comment by ipsi-apant** [comment]
> @ipsi-apant you probably have configured IntelliJ to run Gradle with the same JDK as the Project. ![image](https://user-images.githubusercontent.com/2303663/271607243-354a0d4c-f3a7-40fa-8be3-e149cc358c47.png)

Thank you @steve-todorov 🙇‍♂️ 
After configuring project SDK in intellij build settings, gradle build passed using Java 21 (for 8.4-rc-2).

**Comment by josephglanville** [comment]
I can confirm this is the way to go.

I have Java and Kotlin targets configured as Java 21 and project SDK as 21 but I need to set JDK 20 for Gradle `8.4-rc3` to work correctly.

Note for anyone else also trying to do this will Kotlin you will also need to change the Kotlin compiler version to `1.9.20-Beta2`.

**Comment by goafabric** [comment]
For me this approach is definitely not working .. at least not for native images with graalvm.
Refering to the plugin "org.graalvm.buildtools.native"
First and foremost, it has to be upgrade to the latest "0.9.27" which is not a big deal.
But then the plugin is confused with the different jdk handlings and complains about incompatible java classes.

Pretty sure this can also be configured .. somehow 
But honestly all the "configure java21 but java20" is pretty tedious and mind boggling 
and no where self explationary for anyone cheking out the project.

So i hope that with Kotlin supporting Java21 eventually, we will get a real fix, outside the toolchain config

**Comment by artemptushkin** [comment]
If you have an issue with Gradle 8.4:

```
Unknown Kotlin JVM target: 21
```

Then update Kotlin to `1.9.20-Beta2`, no need to have a snapshot gradle wrapper version

**Comment by simon-aira** [comment]
@artemptushkin How would you upgrade the Kotlin that Gradle uses to execute build scripts?

**Comment by ilyas2016** [comment]
> If you have an issue with Gradle 8.4:
> 
> ```
> Unknown Kotlin JVM target: 21
> ```
> 
> Then update Kotlin to `1.9.20-Beta2`, no need to have a snapshot gradle wrapper version

I am using from intellij. how could i upgrade that?

**Comment by artemptushkin** [comment]
> How would you upgrade the Kotlin that Gradle uses to execute build scripts?

I just compiled (in terminal and Idea) a simple **Gradle Kotlin DSL project too** with 2 changes:

* 
<img width="1116" alt="image" src="https://github.com/gradle/gradle/assets/22632802/d1ffa8ae-ef4d-465a-8eee-abf6b3372fa0">
* 
<img width="934" alt="image" src="https://github.com/gradle/gradle/assets/22632802/66efb9bc-6a19-4f90-a714-3e5583f8ac3f">


> I am using from intellij. how could i upgrade that?

If you want to change this setting then it's possible, intellij inherits Gradle properties so it's just worked for me - no need to change the compiler even because this error is about Gradle


<img width="1105" alt="image" src="https://github.com/gradle/gradle/assets/22632802/f12912fd-4c93-4167-924d-7f62d520599d">

**Comment by izogfif** [comment]
@artemptushkin Do these changes allow building a Gradle multi-module project with `buildSrc` folder containing `*.kts` files? If yes, could you please share your (or sample) project with these settings?

**Comment by steklopod** [comment]
Same for me. 

### Was:
```kotlin
        java { sourceCompatibility = JavaVersion.VERSION_20; targetCompatibility = JavaVersion.VERSION_20 }
        withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile>().configureEach { kotlinOptions { jvmTarget = "20" } }
```


### Now:
```kotlin
        java { toolchain { languageVersion.set(JavaLanguageVersion.of(21)) } }
        withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile>().configureEach { kotlinOptions { jvmTarget = "21" } }
```

Everything is working locally and in github actions with gradle `8.4` & kotlin `1.9.20-RC`. And I don't set `vendor.set(?)`.

### Action:
```yml
      - name: 🪜 Set up JDK 21
        uses: actions/setup-java@v3
        with:
          java-version: 21
          distribution: temurin
```

And in docker I use `FROM amazoncorretto:21 `

**Comment by artemptushkin** [comment]
@izogfif Yes, I think it will work, `.kts` is the same Kotlin file as others. I'd not share, it's a trivial change that probably won't be needed when Gradle updates.

**Comment by namero999** [comment]
@artemptushkin how do you use that `kotlinVersion`? I assume that is in a toml file but what dependency does it apply to?

**Comment by neowu** [comment]
@namero999 i found this way works,  

buildSrc/build.gradle.kts
```
plugins {
    kotlin("jvm") version "1.9.20-RC"
    `kotlin-dsl`
}

repositories {
    mavenCentral()
    gradlePluginPortal()
}

java {
    toolchain {
        languageVersion.set(JavaLanguageVersion.of(21))
    }
}
```

**Comment by steklopod** [comment]
> @namero999 i found this way works,
> 
> buildSrc/build.gradle.kts
> 
> ```
> plugins {
>     kotlin("jvm") version "1.9.20-RC"
>     `kotlin-dsl`
> }
> 
> repositories {
>     mavenCentral()
>     gradlePluginPortal()
> }
> 
> java {
>     toolchain {
>         languageVersion.set(JavaLanguageVersion.of(21))
>     }
> }
> ```

Do you understand why you need `kotlin-dsl` & `gradlePluginPortal()`? :-) 

For any copy-paster dev: please, try different solution and try not to write extra code that is not needed 🤓

**Comment by namero999** [comment]
It is clear that those lines are unrelated to the question. What was missing from the original answer is to what plugin the version would apply, and how to define it.

**Comment by neowu** [comment]
i thought the question was “how to build kotlin dsl project with only jdk 21 rather than jdk17 to boot up gradle with java 21 toolchain” 

the example snippet above is part of buildSrc script, that’s why it contains kotlin dsl and gradle plugin repo

generally 
`kotlin("jvm") version "1.9.20-RC" `
is used to specify kotlin version for compiling gradle kotlin dsl script , which won’t be needed anymore once gradle released full java 21 support 

for a completed working example, anyone can refer to
https://github.com/neowu/core-ng-project/blob/master/buildSrc/build.gradle.kts

java project gradle config is defined in
https://github.com/neowu/core-ng-project/blob/master/buildSrc/src/main/kotlin/project.gradle.kts

both use java 21 toolchain

**Comment by zq99299** [comment]
> 只要尝试一下，它适用于大多数用例。我们从8.3RC1开始就用jdk21编译运行测试，完全没有问题

It's using Gradle+jdk21 to build the project, not compiling and running Gradle using jdk21

**Comment by Villa25698** [comment]
Does Gradle 8.4 work with Java 21 in conjunction with Spring? If it does, how can we make it work? Every time we test our code, we encounter unintended errors. Please assist with resolving this issue:
Unsupported Java. 
Your build is currently configured to use Java 21.0.1 and Gradle 8.2.1.

Possible solution:
 
Use Java 19 as Gradle JVM: Open Gradle settings 
Open Gradle wrapper settings, change distributionUrl property to use compatible Gradle version and reload the project

**Comment by artemptushkin** [comment]
@Villa25698 I think you should go 19 and ask stackoverflow questions, it's impossible to determine what your issue and there might be so many options

Kotlin will release 1.9.20 soon with proper support and it will be straightforward

**Comment by mauritssilvis** [comment]
> Your build is currently configured to use Java 21.0.1 and Gradle 8.2.1.

@Villa25698 If you want to use Java 21 and want to reduce the number of potential problems, you should update your Gradle version to the latest stable version (8.4 at the time of writing). There are two things to note, though: 

1. Gradle will not officially support running on Java 21 until version 8.5, so you might still encounter problems when using Java 21 with Gradle 8.4.
2. Your problem doesn't have to be "Unsupported Java". Upon encountering problems, Gradle might say the Java version is not supported yet, but the actual problem might be another one. For example, if your Spring Boot application starts a Tomcat server and you terminate your application, Gradle will show the "Unsupported Java" error. Nothing went wrong, though. Similarly, other (actual) issues may be hidden behind the "Unsupported Java" error. Have a closer look at the build output or logs to determine the underlying issue.

**Comment by Villa25698** [comment]
we have the problem that Oracle has removed 19 and 20. they only have 17 and 21

Sendt fra Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by Villa25698** [comment]
that's the problem. it isn't the case we set it up incorrectly. earlier we did the same test without changing nothing and it said Gradle is configured java 21.0.1 to Gradle 8.3. therefore it cannot run

Sendt fra Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by quaff** [comment]
> > Still waiting on Kotlin 1.9.20 for the necessary version support in `JvmTarget`.
> 
> I remember reading somewhere that the next Kotlin version was going to be released in December did they mean 1.10/2.0 or did they mean, 1.9.20? Because waiting until December for Java 21 would be long, I am so hyped 😅

Kotlin 1.9.20 is released now.

**Comment by simon-aira** [comment]
I don't see it either [on Github](https://github.com/JetBrains/kotlin/releases) or [on Kotlins release page](https://kotlinlang.org/docs/releases.html)?

**Comment by osiegmar** [comment]
> I don't see it either [on Github](https://github.com/JetBrains/kotlin/releases) or [on Kotlins release page](https://kotlinlang.org/docs/releases.html)?

It‘s currently being published. See for example:

https://repo1.maven.org/maven2/org/jetbrains/kotlin/kotlin-gradle-plugin/1.9.20/

**Comment by ice1000** [comment]
Wow!

**Comment by johnshajiang** [comment]
> Kotlin 1.9.20 is released now.

Does that mean Gradle will support JDK 21 soon?

**Comment by robaho** [comment]
Why is this even an issue? There should be zero dependencies on the JDK from Gradle - unless you change the Gradle code to specifically use a JDK feature - which should be extremely rare.

This points to a less than great internal design of the Gradle codebase - which is a shame.

**Comment by mauritssilvis** [comment]
> Why is this even an issue? There should be zero dependencies on the JDK from Gradle.

Gradle's Kotlin DSL depends on Kotlin, which in turn relies on a specific JDK version. As such, running Gradle is only fully supported for specific JDK versions (i.e., those that Kotlin supports).

**Comment by robaho** [comment]
> > Why is this even an issue? There should be zero dependencies on the JDK from Gradle.
> 
> Gradle's Kotlin DSL depends on Kotlin, which in turn relies on a specific JDK version. As such, running Gradle is only fully supported for specific JDK versions (i.e., those that Kotlin supports).

That is fine - so run Gradle with the required JDK - but have Gradle support whatever language version the build file specifies - there should be no dependency there. But that is not what happens according to:

https://github.com/oracle/javavscode/issues/29

Still, core infra that relies on a subsystem with similar strict dependencies is not the best. The same thing applies to the Kotlin DSL - there should be no reason that shouldn't run with the latest JDK - only if the DSL is changed to support some later feature should it ever be an issue.

It's like developers forgot the forwards compatibility mantra of Java. You've always been able to run compliant Java code on a future JDK release without changes.

**Comment by lion7** [comment]
You know that what you are asking is exactly what is described in this issue? You can already use the toolchains feature to compile and test your own code using JDK 21, it's just that you cannot yet run Gradle itself with JDK 21. I have lots of projects that are already on JDK 21 and those build just fine with Gradle 8.4 (and also with earlier versions). I just use JDK 20 to run Gradle itself.

**Comment by robaho** [comment]
I am sure it’s not gradle’s fault but the gradle language server extension is barging when the compiled class files are 65. 

It is probably MS fault - but still. Clearly the ecosystem doesn’t understand javas forward principles. This should never have been an issue.

**Comment by ice1000** [comment]
@robaho if you read this issue there's a link above on youtrack. Basically, kotlin has an enum with jdk versions, and if it doesn't find a suitable one, it throws an exception. Kotlin team fixed this bug but it takes like two to three months to release the fix.

**Comment by DevDengChao** [comment]
> Kotlin team fixed this bug but it takes like two to three months to release the fix.

Wierd 🤣

**Comment by robaho** [comment]
> @robaho if you read this issue there's a link above on youtrack. Basically, kotlin has an enum with jdk versions, and if it doesn't find a suitable one, it throws an exception. Kotlin team fixed this bug but it takes like two to three months to release the fix.

That is my point. Given the current Java release cycle it is suboptimal for Gradle to couple itself to systems that do no adhere to Javas forward compatibility principles. Drop Kotlin and move on. Or at minimum let the Groovy and Kotlin versions diverge.

**Comment by ice1000** [comment]
> Drop Kotlin and move on

Like rewrite all these gigantic Kotlin codebase? I think rewriting Gradle completely without asm but with new JDK classfile API and without Kotlin/Groovy would be cheaper than that.

**Comment by steve-todorov** [comment]
Guys, this is becoming an offtopic discussion. In a perfect world everything would be perfect, but the reality is we do not live in a perfect world. Deal with it, open a new ticket with your suggestions for improvements or move on to another tool that better fits your needs.

**Comment by robaho** [comment]
How is suggesting splitting gradle into two versions to address issues like this off topic?

it has to be very rare that a package needs both Kotlin and Groovy DSL support.

**Comment by Griefed** [comment]
> How is suggesting splitting gradle into two versions to address issues like this off topic?
> 
> it has to be very rare that a package needs both Kotlin and Groovy DSL support.

It is offtopic because "splitting gradle into two versions" and "rare that a package needs both Kotlin and Groovy DSL" diverge from "Support running with Java 21".

It's an entirely different discussion to be had. I suggest you create another issue and continue the discussion from there, whilst referencing the discussion in this issue here.

**Comment by robaho** [comment]
Filed #26944. Kotlin fanboys feel free to downvote.

**Comment by octylFractal** [comment]
Fixed by #26913

**Comment by 257** [comment]
> Fixed by #26913

```java
Unsupported Java. 
Your build is currently configured to use Java 21.0.1 and Gradle 8.5-rc-1.
```

**Comment by 257** [comment]
@octylFractal you want me to open a new issue?

**Comment by cocorossello** [comment]
> > Fixed by #26913
> 
> ```java
> Unsupported Java. 
> Your build is currently configured to use Java 21.0.1 and Gradle 8.5-rc-1.
> ```

This is intellij that's doing this. Upgrade to 2023.3 beta and you won't see it again

**Comment by 257** [comment]
i'm using android-studio and it seems to be behind `2023.2.1.12`

**Comment by johnshajiang** [comment]
When will be 8.5 GA available?

**Comment by ipsi-apant** [comment]
> > > Fixed by #26913
> > 
> > 
> > ```java
> > Unsupported Java. 
> > Your build is currently configured to use Java 21.0.1 and Gradle 8.5-rc-1.
> > ```
> 
> This is intellij that's doing this. Upgrade to 2023.3 beta and you won't see it again

Yeah this error is from intellij https://youtrack.jetbrains.com/issue/IDEA-326038 . A extremely annoying.
They are using Gradle's compatibility list for validation. https://docs.gradle.org/current/userguide/compatibility.html

Once, gradle updates compatibility list, I think it should work in intellij.

**Comment by ipsi-apant** [comment]
> > Fixed by #26913
> 
> ```java
> Unsupported Java. 
> Your build is currently configured to use Java 21.0.1 and Gradle 8.5-rc-1.
> ```

In intellij settings, if we set JVM 20 for Gradle and project on Java 21. It works.

![image](https://github.com/gradle/gradle/assets/127072273/b5aec580-3d9b-46a2-ab71-824b24e1f2d3)

**Comment by prdoyle** [comment]
@ipsi-apant - when I set it to use Java 20 as you indicated, I get this error:

```
Cause: error: invalid source release: 21
```

Is the language level limited to the Java version used to run Gradle itself?

**Comment by ipsi-apant** [comment]
> @ipsi-apant - when I set it to use Java 20 as you indicated, I get this error:
> 
> ```
> Cause: error: invalid source release: 21
> ```
> 
> Is the language level limited to the Java version used to run Gradle itself?

This was an interim solution until gradle 8.4. After release of Gradle 8.5 and updating intellij to the latest I am not seeing this error.
There was a ticket in YouTrack (can't recall) which described, for gradle based projects, intellij relies on Gradle's Java compatibility matrix.
